Add CI workflow for running doctests

elliotwutingfeng opened the pull request:
Addresses #9 

- Added CI workflow for testing on windows, macos and ubuntu

I found that all tests were passing except on Ubuntu due to mismatched outputs, you can check the GitHub Actions [logs](https://github.com/elliotwutingfeng/dhash/actions/runs/3275505315) for more details.

elliotwutingfeng committed:
Add standard Python gitignore

elliotwutingfeng committed:
Add CI workflow

elliotwutingfeng committed:
Exit 1 if test fails

elliotwutingfeng committed:
Print error output

elliotwutingfeng committed:
Disable fail fast

elliotwutingfeng committed:
Fix typo

elliotwutingfeng committed:
Fix doctest

benhoyt commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,29 @@
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest]
+        python-version: ['2.7', '3.7', '3.8', '3.9', '3.10']
```
Let's drop 3.7 at this point, just to keep the matrix slightly smaller.

benhoyt commented on .gitignore:
```diff
@@ -1,5 +1,162 @@
-__pycache__
-*.pyc
+# Byte-compiled / optimized / DLL files
```
I don't love these "everything" `.gitignore` files, let's just limit it to what we actually need for this project: perhaps all we need is `*.pyc`, `__pycache__/` for now?

elliotwutingfeng replied:
Reverted!

benhoyt commented on dhash.py:
```diff
@@ -256,7 +256,9 @@ def load_image(filename):
         import doctest
-        doctest.testmod()
+        test_results = doctest.testmod()
+        if test_results.failed:
```
The docs seem to indicate it returns a tuple of `failure_count, test_count`. Maybe it's a namedtuple in the implementation, but it doesn't seem like we should rely on this. I think we should change to:

```python
failure_count, _ = doctest.testmod()
if failure_count:
    sys.exit(1)
```

elliotwutingfeng replied:
It is a named tuple, though this approach works too.

benhoyt reviewed (commented):
Thanks for this! Just a couple of minor comments.

Do you want to address the `ANTIALIAS` warning issue from #9 as well?

benhoyt commented:
Do you have any suggestions for how to fix the test failures on Linux? It's odd that the numbers are off by one in some cases, given that it's the same architecture and Pillow/Wand versions. My guess is floating point rounding issues?

elliotwutingfeng committed:
Remove Python 3.7

elliotwutingfeng committed:
Revert gitignore

elliotwutingfeng committed:
Use failure count

elliotwutingfeng committed:
Replace ANTIALIAS with Resampling.LANCZOS

See https://pillow.readthedocs.io/en/stable/reference/Image.html\#PIL.Image.ANTIALIAS

elliotwutingfeng commented:
> Thanks for this! Just a couple of minor comments.
> 
> Do you want to address the `ANTIALIAS` warning issue from #9 as well?

I've replaced it with `PIL.Image.Resampling.LANCZOS`. Test results are the same as before.

elliotwutingfeng committed:
Install libmagickwand-dev for Linux

elliotwutingfeng committed:
Use sudo

elliotwutingfeng committed:
Run apt-get update

elliotwutingfeng commented:
> Do you have any suggestions for how to fix the test failures on Linux? It's odd that the numbers are off by one in some cases, given that it's the same architecture and Pillow/Wand versions. My guess is floating point rounding issues?

Strangely, the tests pass locally on my Ubuntu x64 and Manjaro x64 virtual machines, on both Python 2 and 3 (3.10.6).

elliotwutingfeng committed:
Revert apt-get

benhoyt committed:
See if using PIL for doctests makes tests same on all platforms

Also fall back to PIL.Image.ANTIALIAS for older PIL versions

benhoyt committed:
Pin Pillow and Wand versions

benhoyt committed:
Okay, don't pin versions then, as it breaks the Pillow install on 2.7

ERROR: Could not find a version that satisfies the requirement Pillow==9.2.0 (from versions: 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7.0, 1.7.1, 1.7.2, 1.7.3, 1.7.4, 1.7.5, 1.7.6, 1.7.7, 1.7.8, 2.0.0, 2.1.0, 2.2.0, 2.2.1, 2.2.2, 2.3.0, 2.3.1, 2.3.2, 2.4.0, 2.5.0, 2.5.1, 2.5.2, 2.5.3, 2.6.0, 2.6.1, 2.6.2, 2.7.0, 2.8.0, 2.8.1, 2.8.2, 2.9.0, 3.0.0, 3.1.0rc1, 3.1.0, 3.1.1, 3.1.2, 3.2.0, 3.3.0, 3.3.1, 3.3.2, 3.3.3, 3.4.0, 3.4.1, 3.4.2, 4.0.0, 4.1.0, 4.1.1, 4.2.0, 4.2.1, 4.3.0, 5.0.0, 5.1.0, 5.2.0, 5.3.0, 5.4.0, 5.4.1, 6.0.0, 6.1.0, 6.2.0, 6.2.1, 6.2.2)
ERROR: No matching distribution found for Pillow==9.2.0

benhoyt committed:
Okay, it's still different on 2.7; allow delta of 1

benhoyt commented:
Thanks, @elliotwutingfeng. I added a few more changes to allow older and new versions of Pillow, and "fixed" the tests.

elliotwutingfeng commented:
I re-ran the old tests with this workflow (https://github.com/elliotwutingfeng/dhash/actions/runs/3280926289).

`ubuntu-latest` (`ubuntu-20.04` as of today) uses an older version of `libmagickwand-dev` than `ubuntu-22.04`, that may be causing the output differences. See below.

https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md#installed-apt-packages
https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md#installed-apt-packages

The old tests are successful on 22.04 just like windows and macos. Python 2 had to be manually installed using apt-get on ubuntu-22.04 as it is deprecated. I think the delta workaround is not necessary.

`ubuntu-latest` will be bumped to `ubuntu-22.04` by around [December 2022](https://github.com/actions/runner-images/issues/6399).
